File: app/src/main/java/mk/webfactory/template/data/storage/StorageHelpers.java
package mk.webfactory.template.data.storage;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

public final class StorageHelpers {

  public static <T> Function<T, Observable<T>> deferSaveIn(final Storage<T> storage) {
    return deferSaveIn(storage, false);
  }

  /**
   * Unlike {@linkplain #deferSaveIn}, the create observable will not be allowed to throw,
   * and instead the object being saved will be returned when an error occurs.
   */
  public static <T> Function<T, Observable<T>> deferSaveInAndIgnoreErrors(final Storage<T> storage) {
    return deferSaveIn(storage, true);
  }

  private static <T> Function<T, Observable<T>> deferSaveIn(final Storage<T> storage,
      final boolean continueOnError) {
    return new Function<T, Observable<T>>() {
      @Override public Observable<T> apply(@NonNull final T t) throws Exception {
        if (continueOnError) {
          return storage.save(t).onErrorReturn(new Function<Throwable, T>() {
            @Override public T apply(@NonNull Throwable throwable) throws Exception {
              return t;
            }
          });
        }
        return storage.save(t);
      }
    };
  }

  private StorageHelpers() {
  }
}

File: app/src/main/java/mk/webfactory/template/data/storage/StorageHelpers.kt
package mk.webfactory.template.data.storage

import io.reactivex.Single
import io.reactivex.functions.Function

fun <T> deferSaveIn(storage: Storage<T>): Function<T, Single<T>> {
    return deferSaveIn(storage, false)
}

/**
 * Unlike [deferSaveIn], the create observable will not be allowed to throw,
 * and instead the object being saved will be returned when an error occurs.
 */
fun <T> deferSaveInAndIgnoreErrors(storage: Storage<T>): Function<T, Single<T>> {
    return deferSaveIn(storage, true)
}

private fun <T> deferSaveIn(
    storage: Storage<T>,
    continueOnError: Boolean
): Function<T, Single<T>> {
    return Function { t ->
        if (continueOnError) {
            storage.save(t).onErrorReturn { t }
        } else storage.save(t)
    }
}